test/apps/iodemo/ucx_wrapper.cc

```diff
@@ -687,28 +722,35 @@ void UcxConnection::accept(ucp_conn_request_h conn_req, UcxCallback *callback)
 
 void UcxConnection::disconnect(UcxCallback *callback)
 {
-    /* establish cb must be destroyed earlier since it accesses
-     * the connection */
-    assert(_establish_cb == NULL);
+    UCX_CONN_LOG << "disconnect, ep is " << _ep;
+
     assert(_disconnect_cb == NULL);
+    _disconnect_cb = callback;
 
-    UCX_CONN_LOG << "destroying, ep is " << _ep;
+    _context.remove_connection(this);
 
-    _disconnect_cb = callback;
-    if (ucs_list_is_empty(&_all_requests)) {
-        ep_close(UCP_EP_CLOSE_MODE_FORCE);
-        _context.move_connection_to_disconnecting(this);
-    } else {
-        cancel_all();
-        ep_close(UCP_EP_CLOSE_MODE_FORCE);
+    if (!is_established()) {
+        assert(_ucx_status == UCS_INPROGRESS);
+        established(UCS_ERR_CANCELED);
     }
+
+    _context.move_connection_to_disconnecting(this);
+
+    cancel_all();
+
+    // close the EP after cancelling all outstanding operations to purge all
+    // requests scheduled on the EP which could wait for the acknowledgments
+    ep_close(UCP_EP_CLOSE_MODE_FORCE);
```

Comment:
but we should not close the endpoint until request list is empty on int3 branch
I think need to move this back to disconnect_progress

Reply:
done